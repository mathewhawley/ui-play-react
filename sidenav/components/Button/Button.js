import React, { PropTypes } from 'react';
import classnames from 'classnames';
import styles from './Button.scss';

const Button = ({
  onClick,
  icon,
  customStyles,
}) => {
  const classes = classnames(styles.base, customStyles, 'material-icons');
  return (
    <button className={classes} onClick={onClick}>{icon}</button>
  );
};

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  icon: PropTypes.string.isRequired,
  customStyles: PropTypes.string,
};

export default Button;
