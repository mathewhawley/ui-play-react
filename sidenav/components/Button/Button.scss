.base {
  min-width: 48px;
  min-height: 48px;
}
