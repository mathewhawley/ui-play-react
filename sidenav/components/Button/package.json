{
  "private": true,
  "name": "Button",
  "main": "./Button.js"
}
