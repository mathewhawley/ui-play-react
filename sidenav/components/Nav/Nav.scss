@import '../../styles/variables';

.base {
  overflow: hidden;
  pointer-events: none;
  position: fixed;

  &::before {
    content: '';
    background-color: rgba(0, 0, 0, 0.4);
    position: absolute;
    transition: opacity 150ms ease;
    opacity: 0;
    will-change: opacity;
  }

  &,
  &::before {
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
}

.panel {
  background-color: $white;
  width: 90%;
  max-width: 350px;
  height: 100%;
  position: relative;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.4);
  will-change: transform;
  transform: translateX(-102%);
  transition: transform 250ms ease;
  color: $dark-grey;
}

.header {
  height: 20vh;
  max-height: 200px;
  min-height: 150px;
  background-color: $dark-grey;
  padding: $global-padding;
  color: $white;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.title {
  font-weight: 300;
  margin: 0;
  letter-spacing: 0.01em;
}

.body {
  height: 100%;
  overflow-y: auto;
}

.close {
  color: $white;
  align-self: flex-start;
  margin-top: #{($global-padding - 6) * -1};
  margin-left: #{($global-padding - 6) * -1};
  order: -1;
}

/*
 * States
 * ========================================== */

.base.active {
  pointer-events: auto;

  .panel {
    transform: none;
  }

  &::before {
    opacity: 1;
  }
}
