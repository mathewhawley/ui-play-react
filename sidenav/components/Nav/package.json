{
  "private": true,
  "name": "Nav",
  "main": "./Nav.js"
}
