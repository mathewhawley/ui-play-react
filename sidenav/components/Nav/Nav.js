import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import Button from '../Button';
import styles from './Nav.scss';

class Nav extends Component {
  constructor() {
    super();
    this.navEl = null;
    this.onOutsideClick = this.onOutsideClick.bind(this);
  }

  onOutsideClick({ target }) {
    const { active, toggle } = this.props;
    if (!active || target !== this.navEl) {
      return;
    }
    toggle();
  }

  render() {
    const { active, toggle } = this.props;
    const classes = classnames({
      [styles.base]: true,
      [styles.active]: active,
    });

    return (
      <div className={classes} ref={(node) => this.navEl = node} onClick={this.onOutsideClick}>
        <nav className={styles.panel}>
          <Button icon='close' onClick={toggle} />
        </nav>
      </div>
    );
  }
}

Nav.propTypes = {
  toggle: PropTypes.func.isRequired,
  active: PropTypes.bool,
};

export default Nav;
