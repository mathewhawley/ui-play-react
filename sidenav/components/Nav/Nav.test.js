/* eslint-env jest */

import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Nav from './Nav';
import Button from '../Button';

describe('<Nav />', () => {
  let wrapper;
  let clickHandler;

  beforeEach(() => {
    clickHandler = jest.fn();
    wrapper = shallow(<Nav active={false} toggle={clickHandler} />);
  });

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('renders correctly', () => {
    const tree = renderer.create(
      <Nav active={false} toggle={clickHandler} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call click handler when close button clicked', () => {
    wrapper.find(Button).simulate('click');
    expect(clickHandler).toHaveBeenCalled();
  });

  it('should call click handler when click event is outside the panel', () => {
    const wrapper = mount(<Nav active={true} toggle={clickHandler} />);
    const target = wrapper.instance().navEl;
    wrapper.simulate('click', { target });
    expect(clickHandler).toHaveBeenCalled();
  });

  it('should not call click handler when panel clicked', () => {
    const target = wrapper.find('nav');
    wrapper.simulate('click', { target });
    expect(clickHandler).not.toHaveBeenCalled();
  });
});
