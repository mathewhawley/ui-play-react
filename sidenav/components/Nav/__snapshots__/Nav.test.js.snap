exports[`<Nav /> renders correctly 1`] = `
<div
  className="base"
  onClick={[Function]}>
  <aside
    className="panel">
    <header
      className="header">
      <h1
        className="title">
        Navigation
      </h1>
      <button
        className="base close material-icons"
        onClick={[Function]}>
        close
      </button>
    </header>
    <nav
      className="body">
      <ul>
        <li>
          <a
            href="#">
            Euismod Ornare Elit
          </a>
        </li>
        <li>
          <a
            href="#">
            Euismod Ornare Elit
          </a>
        </li>
        <li>
          <a
            href="#">
            Euismod Ornare Elit
          </a>
        </li>
        <li>
          <a
            href="#">
            Euismod Ornare Elit
          </a>
        </li>
        <li>
          <a
            href="#">
            Euismod Ornare Elit
          </a>
        </li>
        <li>
          <a
            href="#">
            Euismod Ornare Elit
          </a>
        </li>
      </ul>
    </nav>
  </aside>
</div>
`;
